### PR TITLE
[SPARK-25696] The storage memory displayed on spark Application UI is…

### PR DESCRIPTION
**Pull request by httfighter**
… incorrect.

## What changes were proposed in this pull request?
In the reported heartbeat information, the unit of the memory data is bytes, which is converted by the formatBytes() function in the utils.js file before being displayed in the interface. The cardinality of the unit conversion in the formatBytes function is 1000, which should be 1024. 
Change the cardinality of the unit conversion in the formatBytes function to 1024.

## How was this patch tested?
 manual tests

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

**Comment by wangyum** [comment]
cc @ajbozarth

**Review by ajbozarth — APPROVED** [review]
This is correct, the Scala function this replaced used shifts instead of multiplication so it should have been 1024 from the start. Sorry I missed it in my old review. LGTM

**Comment by httfighter** [comment]
It's ok. @ajbozarth

**Comment by wangyum** [comment]
cc @srowen Cloud we merge this PR?

**Comment by srowen** [comment]
Heh this change is actually narrowly wrong, but, is correct and needs to be larger in a broader sense.

1000 is the right factor for KB = kilobytes. 1024 is the right factor for KiB = kibibytes. Spark does generally use the factor 1024 everywhere internally (except one tiny usage I just found). But, all the log and UI messages say "kb" and "mb" and so on. This is incorrect.

If you're willing, this change should really be about, at least, making this change and then updating all usages in the UI to indicate units like "KiB", "MiB", "GiB". What say?

**Comment by srowen** [comment]
@wangyum @httfighter what do you think?

**Comment by wangyum** [comment]
@srowen Thanks for ping me. I agree with you. If @httfighter willing, updating all usages in the UI to indicate units like `KiB`, `MiB`, `GiB`. @httfighter you can check details here: https://en.wikipedia.org/wiki/Kilobyte
https://en.wikipedia.org/wiki/Kibibyte

**Comment by httfighter** [comment]
@srowen Thank you for your review. I agree with you, and I will make changes in the near future.
@wangyum Thank you for your help.

**Comment by httfighter** [comment]
@srowen @ajbozarth I am not sure about some things, can you give me some advice? In the process of modification, I have a question. In Spark, whether M and MB represent MiB.  Spark does not use the unit of kilobytes to convert between numbers.
  private static final ImmutableMap<String, ByteUnit> byteSuffixes =
    ImmutableMap.<String, ByteUnit>builder()
      .put("b", ByteUnit.BYTE)
      .put("k", ByteUnit.KiB)
      .put("kb", ByteUnit.KiB)
      .put("m", ByteUnit.MiB)
      .put("mb", ByteUnit.MiB)
      .put("g", ByteUnit.GiB)
      .put("gb", ByteUnit.GiB)
      .put("t", ByteUnit.TiB)
      .put("tb", ByteUnit.TiB)
      .put("p", ByteUnit.PiB)
      .put("pb", ByteUnit.PiB)
      .build();
<tr>
  <td><code>spark.kryoserializer.buffer</code></td>
  <td>64k</td>
  <td>
    Initial size of Kryo's serialization buffer, in KiB unless otherwise specified. 
    Note that there will be one buffer <i>per core</i> on each worker. This buffer will grow up to
    <code>spark.kryoserializer.buffer.max</code> if needed.
  </td>
</tr>

 If this is the case, can we only guarantee the uniform use of 1024 for digital conversion, no changes to the  unit displays in log, UI, comments and configured messages. Otherwise, we need to modify all the UI, log, comments, and configuration information to ensure consistency, there is no guarantee that all can be modified, and there will be no problems after the modification.

**Comment by srowen** [comment]
I believe most of the Spark code really uses units like KiB, multiples of 1024, as you see here. However it's often referred to as KB, incorrectly, and so on. It seems like the UI is the only place that uses 1000 and KB, which is consistent in itself, but should really be 1024/KiB. that change to the UI seems fine. I think there are a few comments and docs in the code that really also should change to say KiB and so on, and we can make those changes here, but it's OK to focus on making the UI based on KiB.

**Comment by httfighter** [comment]
@srowen OK. Thank you very much for your advice.

**Comment by httfighter** [comment]
@srowen@ajbozarth I have added the changes, could you help me review the code? Thank you very much.

**Comment by srowen** [comment]
Yeah, there are going to be several more tests that fail because they are expecting a string like 'KB'. Hopefully easy to fix.
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99226/testReport/

**Comment by srowen** [comment]
Ping @httfighter

**Comment by httfighter** [comment]
@srowen Sorry, I just saw your message. I am a little busy on weekdays. I will try to modify the test cases in recent days.

**Comment by wangyum** [comment]
retest this please

**Comment by wangyum** [comment]
retest this please

**Review by srowen — CHANGES_REQUESTED** [review]
We also need to change some docs. Among the .md files under docs/, there are several references to values being interpreted as "KB" "MB" or "GB" which should change. I use the regex `\s[EPTGMK]B\W` to find them. Likewise there are several log statements and comments in the code that could change. I'm not worried about line comments, but log statements, javadoc should ideally be updated too.

It's probably another 50 occurrences... if you're up for it great, or else I can open a PR against this one with the additional changes that I think are needed.

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 1]
```diff
@@ -1164,17 +1164,17 @@ private[spark] object Utils extends Logging {
     } else {
       val (value, unit) = {
         if (size >= 2 * EB) {
-          (BigDecimal(size) / EB, "EB")
+          (BigDecimal(size) / EB, "EiB")
```
[the remark this answers is not in the document]
Comment (httfighter) [reply core/src/main/scala/org/apache/spark/util/Utils.scala]: OK! I have submitted a modification.

**Comment by httfighter** [comment]
@srowen Yes. I agree with you! These places should be consistent, otherwise it is easy to be confused. I will try to modify log statements and docs.  Should I modify it in this PR or a new one?

**Comment by srowen** [comment]
Add to this PR. The change goes logically together.

**Comment by httfighter** [comment]
retest this please

**Comment by wangyum** [comment]
retest this please

File: core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala [hunk 2]
```diff
@@ -62,14 +62,14 @@ class KryoSerializer(conf: SparkConf)
 
   if (bufferSizeKb >= ByteUnit.GiB.toKiB(2)) {
     throw new IllegalArgumentException("spark.kryoserializer.buffer must be less than " +
-      s"2048 mb, got: + ${ByteUnit.KiB.toMiB(bufferSizeKb)} mb.")
+      s"2048 mib, got: + ${ByteUnit.KiB.toMiB(bufferSizeKb)} mib.")
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala]: Nit: mib -> MiB. (I know it was 'mb' before but that's not really the abbreviation for 1 million bytes)

File: docs/sql-programming-guide.md [hunk 3]
```diff
@@ -4,10 +4,15 @@ displayTitle: Spark SQL, DataFrames and Datasets Guide
 title: Spark SQL and DataFrames
 ---
 
+* This will become a table of contents (this text will be scraped).
```
Comment (srowen) [inline docs/sql-programming-guide.md]: The change to this file looks unrelated. Could you revert it?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/NullExpressionsSuite.scala [hunk 4]
```diff
@@ -160,7 +160,7 @@ class NullExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
     checkEvaluation(AtLeastNNonNulls(4, nullOnly), false, EmptyRow)
   }
 
-  test("Coalesce should not throw 64kb exception") {
+  test("Coalesce should not throw 64kib exception") {
```
Comment (srowen) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/NullExpressionsSuite.scala]: Nit: 64 KiB

**Review by srowen — CHANGES_REQUESTED** [review]
A few nits, and one file that seems to have been added inadvertently, but otherwise looking good. That's a very comprehensive change, thank you. Some of the changes in comments were arguably not wrong, but, I think it's healthy to try to talk in KiB etc consistently as that is really how most of the code is written. I assume this is basically every single occurrence of the regex in the code!

Fortunately the change is cosmetic and not functional so it should also be safe, as well as more accurate.

**Comment by httfighter** [comment]
retest this please

**Comment by heary-cao** [comment]
retest this please

**Review by srowen — APPROVED** [review]
Looks good to me, and fixes a long-standing inconsistency. Pinging, eh, @HyukjinKwon ? @felixcheung because you commented? maybe @vanzin ? It's just a cosmetic change rather than functional, but, wondered if anyone has any concerns about changing numbers in the UI and log statements.

**Review by felixcheung — APPROVED** [review]
LGTM. what about config values? I thought we had in the form of "100mb"?

**Comment by srowen** [comment]
Fortunately the syntax is "100m", which has always meant "100 * 1024 * 1024" or "100 MiB"

**Comment by HyukjinKwon** [comment]
Looks okay to me too. UI change requires screenshots of UI tho strictly.

**Comment by srowen** [comment]
In this case the change is simpler to understand in prose, I think; "100 KB" becomes "97.6 KiB", etc.
